hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveSchemaUtils.java

```diff
@@ -0,0 +1,691 @@
+          oldSchema.getType(), newSchema.getType());
+      return true;
+    }
+
+    switch (newSchema.getType()) {
+      case BOOLEAN:
+      case BYTES:
+      case DOUBLE:
+      case FLOAT:
+      case INT:
+      case LONG:
+      case STRING:
+      case FIXED:
+        return false;
+      case ENUM:
+        return oldSchema.getEnumSymbols().size() != newSchema.getEnumSymbols().size();
+
+      case RECORD:
+        return isRecordSchemaEvolved(oldSchema, newSchema);
+
+      case MAP:
+        return isSchemaEvolved(oldSchema.getValueType(), newSchema.getValueType());
+
+      case ARRAY:
+        return isSchemaEvolved(oldSchema.getElementType(), newSchema.getElementType());
+
+      case UNION:
+        boolean isBothNullableType =
+            AvroSchemaUtil.isOptionSchema(oldSchema) && AvroSchemaUtil.isOptionSchema(newSchema);
+
+        if (!isBothNullableType) {
+          throw new IllegalArgumentException(
+              "We do not support UNION type " + "except nullable field in schema: " + newSchema.toString(true));
+        }
+
+        Schema oldNonNullType = AvroSchemaUtil.fromOption(oldSchema);
+        Schema newNonNullType = AvroSchemaUtil.fromOption(newSchema);
+
+        return isSchemaEvolved(oldNonNullType, newNonNullType);
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported Avro type " + newSchema.getType() + " in new schema: " + newSchema.toString(true));
+    }
+  }
+
+  private static boolean isPartitioned(@Nonnull Table tableOrView) {
+    Preconditions.checkNotNull(tableOrView, "tableOrView cannot be null");
+
+    List<FieldSchema> partitionColumns = getPartitionCols(tableOrView);
+
+    if (partitionColumns == null) {
+      return false;
+    }
+
+    return partitionColumns.size() != 0;
+  }
+
+  private static List<FieldSchema> getPartitionCols(@Nonnull Table tableOrView) {
+    Preconditions.checkNotNull(tableOrView, "tableOrView cannot be null");
+
+    List<FieldSchema> partKeys = tableOrView.getPartitionKeys();
+    if (partKeys == null) {
+      partKeys = new ArrayList<>();
+      tableOrView.setPartitionKeys(partKeys);
+    }
+
+    return partKeys;
+  }
+
+  private static void mergeAndAppendField(Schema.Field sourceField, Schema.Field metaDataField,
+      SchemaBuilder.FieldAssembler<Schema> fieldAssembler) {
+    Schema.Type actualMetaDataFieldType = extractActualTypeIfFieldIsNullableType(metaDataField.schema()).getType();
+    boolean isSameType = sourceField.schema().getType().equals(actualMetaDataFieldType);
+    boolean isMetaDataFieldEnumType = metaDataField.schema().getType().equals(Schema.Type.ENUM);
+    boolean useMetaDataType = isSameType || isMetaDataFieldEnumType;
+
+    Object defaultValue = metaDataField.defaultVal();
+    if (defaultValue != null) {
+      fieldAssembler.name(metaDataField.name()).doc(metaDataField.doc())
+          .type(useMetaDataType ? metaDataField.schema() : sourceField.schema()).withDefault(defaultValue);
```

Comment:
Take `fieldAssembler.name(metaDataField.name()).doc(metaDataField.doc()).type(useMetaDataType ? metaDataField.schema() : sourceField.schema())` out of the condition and only adjust the default value inside the `if` statement. Is the default `noDefault()`? Then maybe no need for the `noDefault()` branch, but please double check.